File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    using pii = pair<int,int>;
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> dict;
        for(auto& i:nums) dict[i]++;
        auto cust_comp = [](const pii& a,const pii& b){
            return a.second>b.second;
        };
        priority_queue<pii,vector<pii>,decltype(cust_comp)> pq(cust_comp);
        for(auto& i:dict){
            pq.push({i.first,i.second});
            if(pq.size()>k) pq.pop();
        }
        vector<int> res;
        while(pq.size()){
            res.push_back(pq.top().first);
            pq.pop();
        }
        return res;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
from collections import defaultdict
from heapq import heappush,heappop

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = defaultdict(int)
        for val in nums: 
            freq[val]+=1 
        min_heap = []
        for key,fr in freq.items(): 
            heappush(min_heap,(fr,key))
            if len(min_heap)>k: 
                heappop(min_heap)
        res = []
        while min_heap:
            res.append(heappop(min_heap)[1])
        return res